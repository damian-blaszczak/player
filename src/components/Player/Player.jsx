import React, { useState } from 'react';
import Topbar from '../Topbar/Topbar';
import styled from 'styled-components';
import useAudioPlayer from '../../lib/hooks/useAudioPlayer';
import Track from './Track/Track';
import { useQuery } from 'react-query';
import { getTracksOnce } from '../../lib/api/getTracks';
import Loader from '../Loader/Loader';
import TabsComponent from '../TabsComponent/TabsComponent';
import SampleStyledParagraph from '../SampleStyledParagraph/SampleStyledParagraph';

const PlayerContainer = styled.div`
  background: rgb(83,64,84);
  background: linear-gradient(180deg, rgba(83,64,84,1) 0%, rgba(41,29,60,1) 40%, rgba(41,29,60,1) 100%);
  min-height: 100vh;
  height: 100%;
  color: #fff;
`;

const Player = () => {
  const [tabIndex, setTabIndex] = useState(1);
  const { curTime, playing, duration, setPlaying, setTime } = useAudioPlayer();
  const { status: tracksStatus, data: tracksList } = useQuery('tracks', getTracksOnce);
  const [trackIndex, setTrackIndex] = useState(1);

  const pageTitle = tracksList && tracksList.find(track => track.id === trackIndex)?.name;

  const handleChange = (_event, newValue) => {
    setTabIndex(newValue)
    if (playing) {
      setPlaying(false);
    }
  };

  const fastForward = () => {
    setTime(curTime + 30);
  }

  const rewind = () => {
    setTime(curTime - 15);
  }

  const nextTrack = () => {
    if (trackIndex < tracksList.length) {
      setTrackIndex(trackIndex + 1);
    } else {
      setTrackIndex(1);
    }
    setPlaying(false);
  };

  const previousTrack = () => {
    if (trackIndex !== 1) {
      setTrackIndex(trackIndex - 1);
    } else {
      setTrackIndex(tracksList.length);
    }
    setPlaying(false);
  };

  const preloadImages = () => tracksList && tracksList.forEach(track => {
    new Image().src = track.image;
  });

  return (
    <PlayerContainer>
      {tracksStatus === 'loading'
        ? <Loader />
        : (
          <>
            <Topbar title={pageTitle} />
            {tracksList.map(track => trackIndex === track.id && (
              <TabsComponent
                key={track.id}
                tabIndex={tabIndex}
                handleChange={handleChange}
                tabs={['OPIS', 'SŁUCHAJ', 'ROZDZIAŁY']}
                tabPanels={[
                  <SampleStyledParagraph data-testid="trackDescription">Tutaj znajduje się opis utworu {track.name}.</SampleStyledParagraph>,
                    <React.Fragment>
                      <audio id="audio">
                        <source src={track.track} />
                        Your browser does not support the <code>audio</code> element.
                      </audio>
                      <Track
                        trackName={track.name}
                        trackImage={track.image}
                        fastForward={fastForward}
                        rewind={rewind}
                        playing={playing}
                        setPlaying={setPlaying}
                        nextTrack={nextTrack}
                        previousTrack={previousTrack}
                        curTime={curTime}
                        duration={duration}
                        preloadImages={preloadImages}
                      />
                    </React.Fragment>,
                  <SampleStyledParagraph>Tutaj znajdują się rozdziały utworu {track.name}.</SampleStyledParagraph>
                ]}
              />
            ))}
          </>
        )
      }
    </PlayerContainer>
  );
};

export default Player;
