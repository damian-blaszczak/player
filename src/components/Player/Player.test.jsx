import React from 'react';
import { render, cleanup, fireEvent } from '@testing-library/react';
import Player from './Player';
import sinon from 'sinon';
import * as reactQuery from 'react-query';
import { MemoryRouter as Router } from 'react-router-dom';

describe('Player', () => {
  const sandbox = sinon.createSandbox();
  beforeAll(() => {
    sandbox.stub(reactQuery, 'useQuery').returns({
      data: [
        {
          id: 1,
          name: 'track1',
          image: 'imageUrl1',
          track: 'mp3track1'
        }, {
          id: 2,
          name: 'track2',
          image: 'imageUrl2',
          track: 'mp3track2'
        }, {
          id: 3,
          name: 'track3',
          image: 'imageUrl3',
          track: 'mp3track3'
        }
      ],
      isLoading: false,
      error: null
    });
    window.HTMLMediaElement.prototype.play = jest.fn();
    window.HTMLMediaElement.prototype.pause = jest.fn();
  });
  afterEach(() => {
    cleanup();
  });

  it('renders correctly', () => {
      const { container } = render(
        <Router>
          <Player />
        </Router>
      );

      expect(container).toMatchSnapshot();
  });

  it('should change state playing', () => {
    // Arrange
    const { queryByText } = render(
      <Router>
        <Player />
      </Router>
    );

    const playButton = queryByText('i_play.svg');
    const pauseButton = queryByText('i_pause.svg');

    // Assert
    expect(pauseButton).toBeNull();
    expect(playButton).toBeTruthy();

    // Act
    fireEvent.click(playButton);

    // Assert
    expect(queryByText('i_pause.svg')).toBeTruthy();
    expect(queryByText('i_play.svg')).toBeNull();
  });

  it('should change state trackIndex', () => {
    // Arrange
    const { queryByTestId, queryByAltText } = render(
      <Router>
        <Player />
      </Router>
    );

    // Assert
    expect(queryByAltText('track1-cover')).toBeTruthy();
    expect(queryByAltText('track2-cover')).toBeNull();

    // Act
    fireEvent.click(queryByTestId('nextTrack'));

    // Assert
    expect(queryByAltText('track1-cover')).toBeNull();
    expect(queryByAltText('track2-cover')).toBeTruthy();

    // Act
    fireEvent.click(queryByTestId('previousTrack'));

    // Assert
    expect(queryByAltText('track1-cover')).toBeTruthy();
    expect(queryByAltText('track2-cover')).toBeNull();
  });

  it('should change current tab', () => {
    // Arrange
    const { queryByText, queryByTestId, queryByAltText } = render(
      <Router>
        <Player />
      </Router>
    );

    // Assert
    expect(queryByAltText('track1-cover')).toBeTruthy();
    expect(queryByTestId('trackDescription')).toBeNull();

    // Act
    fireEvent.click(queryByText('OPIS'));

    // Assert
    expect(queryByAltText('track1-cover')).toBeNull();
    expect(queryByTestId('trackDescription')).toBeTruthy();
  });
});
