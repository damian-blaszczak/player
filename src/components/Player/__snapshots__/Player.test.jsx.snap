// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Player renders correctly 1`] = `
<div>
  <div
    class="sc-fznyAO cEwMaW"
  >
    <div
      class="sc-AxirZ fyWRYB"
    >
      <a
        href="/"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
          focusable="false"
          style="color: rgb(255, 255, 255);"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
          />
        </svg>
      </a>
      <h2
        class="sc-AxjAm ijAwJO"
      >
        track1
      </h2>
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
        focusable="false"
        style="color: rgb(255, 255, 255);"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
        />
      </svg>
    </div>
    <div
      class="MuiTabs-root"
    >
      <div
        class="MuiTabs-scroller MuiTabs-fixed"
        style="overflow: hidden;"
      >
        <div
          class="MuiTabs-flexContainer"
          role="tablist"
        >
          <button
            aria-selected="false"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit MuiTab-fullWidth"
            role="tab"
            tabindex="-1"
            type="button"
          >
            <span
              class="MuiTab-wrapper"
            >
              <span
                class="sc-fzoLsD jqWCJU"
              >
                OPIS
              </span>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
          <button
            aria-selected="true"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit Mui-selected MuiTab-fullWidth"
            role="tab"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiTab-wrapper"
            >
              <span
                class="sc-fzoLsD jqWCJU"
              >
                SŁUCHAJ
              </span>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
          <button
            aria-selected="false"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorInherit MuiTab-fullWidth"
            role="tab"
            tabindex="-1"
            type="button"
          >
            <span
              class="MuiTab-wrapper"
            >
              <span
                class="sc-fzoLsD jqWCJU"
              >
                ROZDZIAŁY
              </span>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
        <span
          class="PrivateTabIndicator-root-2 PrivateTabIndicator-colorSecondary-4 MuiTabs-indicator makeStyles-customTabIndicator-1"
          style="left: 0px; width: 0px;"
        />
      </div>
    </div>
    <div
      class="sc-fzpans kMygFH"
    />
    <div
      class="sc-fzpans kMygFH"
    >
      <audio
        id="audio"
      >
        <source
          src="mp3track1"
        />
        Your browser does not support the 
        <code>
          audio
        </code>
         element.
      </audio>
      <div
        class="sc-AxiKw jSdSnB"
      >
        <img
          alt="track1-cover"
          class="sc-AxhCb gTjuCn"
          src="imageUrl1"
        />
        <div
          class="sc-AxheI gJsONp"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
            data-testid="previousTrack"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
            />
          </svg>
          <h3
            class="sc-AxhUy iuHHVd"
          >
            track1
          </h3>
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
            data-testid="nextTrack"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
            />
          </svg>
        </div>
        <div
          class="sc-Axmtr cLRAwJ"
        >
          <div
            class="sc-AxgMl iKWKjR"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              style="font-size: 4rem;"
              viewBox="0 0 24 24"
            >
              <path
                d="M12.5 8c-2.65 0-5.05.99-6.9 2.6L2 7v9h9l-3.62-3.62c1.39-1.16 3.16-1.88 5.12-1.88 3.54 0 6.55 2.31 7.6 5.5l2.37-.78C21.08 11.03 17.15 8 12.5 8z"
              />
            </svg>
            <span
              class="sc-AxmLO ieIqjG"
            >
              15
            </span>
          </div>
          <svg
            style="width: 100px; height: 100px;"
          >
            i_play.svg
          </svg>
          <div
            class="sc-AxgMl iKWKjR"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              style="font-size: 4rem;"
              viewBox="0 0 24 24"
            >
              <path
                d="M18.4 10.6C16.55 8.99 14.15 8 11.5 8c-4.65 0-8.58 3.03-9.96 7.22L3.9 16c1.05-3.19 4.05-5.5 7.6-5.5 1.95 0 3.73.72 5.12 1.88L13 16h9V7l-3.6 3.6z"
              />
            </svg>
            <span
              class="sc-AxmLO ieIqjG"
            >
              30
            </span>
          </div>
        </div>
        <span
          class="sc-AxmLO ieIqjG"
        >
          NaN:0NaN
           / 
          NaN:0NaN
        </span>
      </div>
    </div>
    <div
      class="sc-fzpans kMygFH"
    />
  </div>
</div>
`;
